### PR TITLE
Allow _.Property / _.MethodCall() / _.IndexerAccess[idx]  shorthand for accessor functions

### PR DESCRIPTION
**Pull request by tboby**
_.foo.bar- > fun x - > x.foo.bar

link to suggestion - https://github.com/fsharp/fslang-suggestions/issues/506
RFC PR : https://github.com/fsharp/fslang-design/pull/710

### Allowed constructs

Treating _.expression as fun x -> x.expression 

```

let a1 : {| Foo : {| Bar : string |}|} -> string = _.Foo.Bar
let a2 : {| Foo : int array |} -> int = _.Foo.[5]
let a3 : {| Foo : int array |} -> int = _.Foo[5]
let a4 : {| Foo : unit -> string |} -> string = _.Foo()
let a5 : {| Foo : int -> {| X : string |} |} -> string = _.Foo(5).X
let a6 = [1] |> List.map _.ToString()
```

Now with correct coloring and autocomplete:
image

And with right-hand-side derivation of the input type:
(not saying I would recommend such style of coding)
image

### CODE REVIEW

**Comment by vzarytovskii** [comment]
I've renamed it and added link to the suggestions, hope you don't mind @tboby

**Comment by tboby** [comment]
I think this is now a working proof of concept that implements the minimum features described in the suggestions issue.

Things I'm pretty sure need improving:

- [ ] The way a conflicting underscore is detected for the diagnostic (performance + check starts with `_` is a safe assumption)
- [ ] Find a better name for the syntax tree object
- [ ] Proper language server features (exposing enough for hover/etc)
- [ ] Improve the walking of the tree to push the arg down into the shorthand
- [ ] Confirm the parser rule is OK (the %prec is what worked, I didn't reason it out)

**Comment by vzarytovskii** [comment]
@tboby please also make sure that the feature is under a language feature flag.

**Comment by vzarytovskii** [comment]
This also needs an RFC

**Comment by abelbraaksma** [comment]
This is pretty darn awesome! If you need some help with RFC, ping me on Slack or @-mention me here.

**Comment by tboby** [comment]
> Hi @tboby, I really like this idea.
> 
> Would you mind if I try to have a look and attempt to revive this PR? Or are you planning to continue on it?

Absolutely, go ahead! I keep meaning to try and continue but haven't found the time.

Use as much of this and/or the RFC as is useful!

**Comment by T-Gro** [comment]
I spent some time trying to support _.prop in pattern matching, if user code supplies its own active pattern to handle it.
However, the needed syntax constructs are not possible when in a match clause, so I am giving up.

(even a regular lambda (fun x -> x.Prop) leads to a parse error, as it is unexpected in matching context.

**Comment by T-Gro** [comment]
Once CI passes, this is ready for review.

File: tests/FSharp.Compiler.ComponentTests/Language/DotLambdaTests.fs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+open System
+let a6 = [1] |> List.map _.ToString()
+        """
+        |> withLangVersionPreview
+        |> typecheck
+        |> shouldSucceed
+        
+    [<Fact>]
+    let ``Nested anonymous unary function shorthands compile`` () =
```
Comment (abonie) [inline tests/FSharp.Compiler.ComponentTests/Language/DotLambdaTests.fs:74]: Should it compile as the name of the test suggests? Or should it fail to compile as it does in the test?

Comment (abonie) [reply tests/FSharp.Compiler.ComponentTests/Language/DotLambdaTests.fs:74]: Sorry, I only noticed now this is still(/again) a draft

Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Language/DotLambdaTests.fs:74]: It should fail, test is to be updated

File: src/Compiler/pars.fsy [hunk 2]
```diff
@@ -4788,6 +4789,14 @@ argExpr:
         arg }
 
 atomicExpr:
+  | UNDERSCORE DOT atomicExpr %prec dot_lambda
+      { let arg1 = rhs parseState 1
```
Comment (nojaf) [inline src/Compiler/pars.fsy:4793]: Nit, but `mUnderscore` and `mDot` would be better than `arg1` and `arg2`.
`arg3` could be `expr`.

Comment (T-Gro) [reply src/Compiler/pars.fsy:4793]: Good point, will change

**Review by nojaf — COMMENTED** [review]
Could you please add some tests in `fsharp\tests\service\data\SyntaxTree`?

And I guess a test for `tests\FSharp.Compiler.ComponentTests\Signatures\TestCasesForGenerationRoundTrip` might also be interesting. Consider the case: `let x = _.ToString()`.

There doesn't seem to be any recovery story for `_.`.

Consider:
```
_.

()
```

And have you considered the following:

```fsharp
let x = _.ToString() // SynExpr.DotLambda, val x: _arg1: 'a -> string
let y = _.ToString () // SynExpr.App(NonAtomic, DotLambda, Const(Unit)), val y: (unit -> string)

let a = "a"
printfn "%A" (x a)
printfn "%A" (y a)

let b = a.ToString() // SynExpr.App(Atomic, LongIdent, Const(Unit))
let c = a.ToString  () // SynExpr.App(NonAtomic, LongIdent, Const(Unit))
```

Adding the space in `y`, no longer leads to the same `SynExpr.DotLambda`.
While `b` and `c` still lead to `SynExpr.App`.
How do we want to handle that? A certain code formatter now supports having that extra space in applications. If that is not at all allowed for the new dot lambda, it should be made very clear. Because users will report inconsistency bugs when they see this.

And as a closing thought (nothing you want to hear this late in the game):
Are we in the right direction with the syntax tree by having yet another `SynExpr` to capture something with dots? Should we not start unifying`SynExpr.DotGet`, `SynExpr.DotIndexedGet`, `SynExpr.LongIdent` at some point?

**Comment by auduchinok** [comment]
> Are we in the right direction with the syntax tree by having yet another SynExpr to capture something with dots? Should we not start unifying SynExpr.DotGet, SynExpr.DotIndexedGet, SynExpr.LongIdent at some point?

It seems a new union case for `_` in `SynExpr` would work nice, and it could be used as a qualifier in `DotGet` expressions.

**Comment by nojaf** [comment]
> It seems a new union case for _ in SynExpr would work nice, and it could be used as a qualifier in DotGet expressions.

Yes, played around with that in https://github.com/dotnet/fsharp/pull/15548
The drawback of course is that the `SynExpr.Wild` on itself doesn't really mean anything.
Although overall, it does feel a lot closer to the existing constructs.

**Comment by T-Gro** [comment]
@nojaf:

I think that the existing constructs are "undecided" on the granularity as well - doBang, matchBang exist as a more meaningful standalone SynExpr, instead of having a meaning-less (on its own) SynExpr.Bang.
=> I think the case could be made either way here.


For the spaces:
I personally did not want to support spaces here.
The poor-mans explanation of what's allowed following _. (very roughly, but 'explain to a 5 year old' style) is an expression without any spaces. So just a chain of primitive expressions, potentially separated by dots.


If by a different approach it would start to work, I would actually vote for not allowing  `_.ToString ()`

**Comment by T-Gro** [comment]
Agree to the rest, will add the cases you mentioned so that it is visible already in this PR diff on how the tree looks like.

**Comment by nojaf** [comment]
> I think that the existing constructs are "undecided" on the granularity as well - doBang, matchBang exist as a more meaningful standalone SynExpr, instead of having a meaning-less (on its own) SynExpr.Bang.

I guess it is very questionable whether these really had their own right to exist. From a syntax point of view, they can be captured in their non-bang counterpart.

At the same time, the updated index syntax (`arr[0]`) did not receive a special AST node. So I'm pretty sure you can find examples on both sides of this.

> If by a different approach, it would start to work, I would actually vote for not allowing _.ToString ()

I very much would like to see that on record in the RFC. Because this will pop up eventually as a bug report on Fantomas' side.

Should an application with a `SynExpr.DotLambda` produce a warning if it is decided that it is illegal?

**Comment by T-Gro** [comment]
I have added:
- sigdata roundtrip test
- syntaxtree tests for positive and negative scenarios (some of the negative ones do not end up with an DotLambda syntraxtree item eventually)
- New negative tests for incorrect feature usage (puting spaces before "()" or using curried function call)


I am open to suggestions on how to:
- Spot the incorrect feature usage, especially where to handle it best 
- Texting to change, the current messages in the test assertions are just the out of the box messages in case nothing gets changed.

**Comment by T-Gro** [comment]
The opening suggestion would be to analyze the inner synExpr (of the DotLambda node) in CheckExpressions and look for NonAtomic App . If one is found , report it.

Something like:
" Underscore-dot lambda shorthand allows method calls with arguments in parenthesis, not separated by a space from the member name".

**Comment by T-Gro** [comment]
> > I think that the existing constructs are "undecided" on the granularity as well - doBang, matchBang exist as a more meaningful standalone SynExpr, instead of having a meaning-less (on its own) SynExpr.Bang.
> 
> I guess it is very questionable whether these really had their own right to exist. From a syntax point of view, they can be captured in their non-bang counterpart.
> 
> At the same time, the updated index syntax (`arr[0]`) did not receive a special AST node. So I'm pretty sure you can find examples on both sides of this.
> 
> > If by a different approach, it would start to work, I would actually vote for not allowing _.ToString ()
> 
> I very much would like to see that on record in the RFC. Because this will pop up eventually as a bug report on Fantomas' side.
> 
> Should an application with a `SynExpr.DotLambda` produce a warning if it is decided that it is illegal?

I think the technical differentiation will be in not allowing NonAtomic  App nodes (unless someone/something proves me wrong in that, it is just my initial guess).

The error message for end users must of course use a different vocabulary, I am open for suggestions here.

**Comment by vzarytovskii** [comment]
Yeah, we should (for now) disallow spaces and add note  it in pr. Support for spaces and partial application (for untupled method) might be an additional feature.

**Comment by T-Gro** [comment]
The diagnostics for incorrect usages updated and reflected in tests:
- ambigous usage of _
- using non-atomic expressions after _.

**Review by abonie — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**